### PR TITLE
Add warning modal to order cycle with attached schedule general setting form [OFN-11613]

### LINKED ISSUE #11613: Display a warning message when changing the date range of an OC with linked orders
## Description
**- As a:** enterprise user or super admin
**- On page:** `/admin/order_cycles`
**- I want to be able to do:** 

To create a new order cycle (OC), some users change the date range of an old one instead of duplicating it first.
This has hard consequences: producers bringing too much food on the delivery day and a big mess with orders.

When changing the date range of an OC **which has linked orders**, display a popup, for instance: "Orders are linked to this order cycle. If you wish to create a new order cycle, it is recommended to duplicate the order cycle first and then change the dates."

With two buttons as:
"Proceed anyway" -- "Cancel"

Proceed anyway => the saving process goes according to what the user has selected (saving and staying on the form, or saving and going back to the list)

Cancel should move the user back to the list (where the duplicate button is visible)

Here's the mock-up for guidance:
image

Here's the Figma link to the design:
https://www.figma.com/file/8iCdevuefE4hRIbGyBhWHX/Nicholas-Pushes?type=design&node-id=1%3A3&mode=design&t=seDLvGHNaCuVMQLs-1

Many thanks to @NickIkenma for the design!!

## Acceptance Criteria & Tests

1. Log in as an enterprise
2. Edit a currently active or past OC (make sure the OC has orders linked)
3. See the message
4. Create an OC (past, current or future) see that you can edit it without seeing the warning
5. Test you can cancel and be redirected to the OC list or proceed according to your saving button choice (staying on the form or be redirected to the list

Repeat as super admin.

This needs to be covered by an automated test.

### PR DESCRIPTION
**Pull request by wandji20**
#### What? Why?

- Closes #11613

To enable a warning modal when editing an order cycle with a linked schedule, I added
- A link modal in the order cycle edit template (trigger activated in save bar when date time range value changes)
- A  stimulus controller to monitor changes in the date time range values and display appropriate save bar buttons.

#### What should we test?
- Editing a not-linked schedule order cycle should work as before
- In case an order cycle has a linked schedule, if a change is detected in the date time range values (orders open at and close at) the user will see a warning modal on an attempt to 'save', 'save and next', or 'save and back to list'.
However, on confirming the modal, the corresponding user action will be triggered.

- Visit ... page.
 http://localhost:3000/admin/order_cycles/:id/edit

#### Release notes

N/A

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [ ] Technical changes only

Add warning modal to order cycle with attached schedule general setting form

#### Dependencies
N/A

#### Documentation updates
N/A

### CODE REVIEW

**Comment by RachL** [comment]
@wandji20 same here, it looks like the PR needs a rebase

File: app/views/admin/order_cycles/edit.html.haml [hunk 1]
```diff
@@ -15,19 +15,46 @@
   = t :edit_order_cycle
 
 - ng_controller = @order_cycle.simple? ? 'AdminSimpleEditOrderCycleCtrl' : 'AdminEditOrderCycleCtrl'
+- has_scheduled_order = @order_cycle.schedules.exists?
```
Comment (dacook) [inline app/views/admin/order_cycles/edit.html.haml:18]: According to the issue, I think this should be checking for linked orders, not schedules. 

> When changing the date range of an OC **which has linked orders**

There is a DB field `orders.order_cycle_id`. So I think we can add a relationship to `order.rb`: `has_many :orders` (I'm not sure why this isn't there already).

Then it should be easy to check if `@order_cycle.orders.exists?`

File: app/webpacker/controllers/order_cycle_controller.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    this.modalConfirmTargets.forEach(ele => {
+      if (e.target.getAttribute('data-target') === ele.getAttribute('data-request')) {
+        ele.style.display = 'unset';
+      } else {
+        ele.style.display = 'none';
+      }
+    });
+  }
+
+  updateCallback(data) {
```
Comment (rioug) [inline app/webpacker/controllers/order_cycle_controller.js:42]: I see we are using this as a way to update `initValsValue` when the order cycle is update via angular :+1:  nice trick to allow stimmulus and angular to communicate !

Could you add a comment to explain why we are using this ?

File: app/webpacker/controllers/order_cycle_controller.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  updateModalConfirmButton(e) {
+    if(!this.hasScheduleValue) return;
+    // Display modal confirm button coresponding to save bar button clicked
+    this.modalConfirmTargets.forEach(ele => {
+      if (e.target.getAttribute('data-target') === ele.getAttribute('data-request')) {
+        ele.style.display = 'unset';
+      } else {
+        ele.style.display = 'none';
+      }
+    });
```
Comment (rioug) [inline app/webpacker/controllers/order_cycle_controller.js:39]: It's a bit hard to follow what's happening, but if understand correctly you are using the `data-target` to display the correct button the modal, right ?
I think you could just pass a parameter to  `updateModalConfirmButton()` to achieve the same thing, and I believe it would make the intent of the function easier to understand. 
Something like : 

```suggestion 
    this.modalConfirmTargets.forEach(ele => {
      if (e.params.request === ele.getAttribute('data-request')) {
        ele.style.display = 'unset';
      } else {
        ele.style.display = 'none';
      }
    });
```
Then you can call it like so :
```
 %input.red{ type: "button", value: t('.save_and_next'), "ng-disabled": "!order_cycle_form.$dirty || order_cycle_form.$invalid", data: { 'action': 'click->modal-link#open click->order-cycle#updateModalConfirmButton', 'order-cycle-request-param': 'saveAndNext'} }
```

File: app/views/admin/order_cycles/edit.html.haml [hunk 4]
```diff
@@ -15,19 +15,46 @@
...
+            = t('admin.order_cycles.edit.linked_schedule_warning_modal.content')
+        %p.modal-actions.justify-end
+          %button.button.secondary#modal-confirm{ type: "button", 'data-action': 'click->modal#close', style: 'display: none;', data: { 'order-cycle-target': 'modalConfirm', request: 'save' }, "ng-click": "submit($event, null)" }
+            = t('admin.order_cycles.edit.linked_schedule_warning_modal.proceed')
+          %button.button.secondary#modal-confirm{ type: "button", 'data-action': 'click->modal#close', style: 'display: none;', data: { 'order-cycle-target': 'modalConfirm', request: 'saveAndNext' }, "ng-click": "submit($event, '#{main_app.admin_order_cycle_incoming_path(@order_cycle)}')" }
+            = t('admin.order_cycles.edit.linked_schedule_warning_modal.proceed')
+          %button.button.secondary#modal-confirm{ type: "button", 'data-action': 'click->modal#close', style: 'display: none;', data: { 'order-cycle-target': 'modalConfirm', request: 'saveAndBack' }, "ng-click": "submit($event, '#{main_app.admin_order_cycles_path}')" }
+            = t('admin.order_cycles.edit.linked_schedule_warning_modal.proceed')
```
Comment (rioug) [inline app/views/admin/order_cycles/edit.html.haml:58]: I am not a big fan of having a mix of stimulus and angular actions on the same button, but can't really think of another way of doing it ?

**Review by rioug — CHANGES_REQUESTED** [review]
Nice one @wandji20 :+1:  , I like that you found a way to use stimulusjs and have angularjs communicate with the stimulus controller.
Ideally we would remove angulajs all together but that's out of scope for this.

I think we can improve your solution a little, could you have a look at my comments  ?

File: app/webpacker/controllers/order_cycle_controller.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    // Attach update callback method
+    window.adminOrderCycleUpdateCallback = this.updateCallback.bind(this);
+  }
+
+  toggleSaveBtns() {
+    if(!this.hasScheduleValue) return;
+
+    // Check that datetime input value has a change
+    const dirty = this.inputTargets.some(ele =>
+      new Date(this.initValsValue[`${ele.name}`]).getTime() !== new Date(ele.value).getTime());
```
Comment (dacook) [inline app/webpacker/controllers/order_cycle_controller.js:18]: Instead of having to pass the current date in as a variable, we can actually ask the DOM by using `ele.defaultValue`. See example here:

https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/webpacker/controllers/bulk_form_controller.js#L163


Then you won't need to add the `initVals` values and the HTML can be cleaner.

File: app/webpacker/css/admin/order_cycles.scss [hunk 6]
```diff
@@ -62,3 +62,25 @@ form.order_cycle {
     }
   }
 }
+
+#linked-schedule-warning-modal {
```
Comment (dacook) [inline app/webpacker/css/admin/order_cycles.scss:66]: Are any of these styles improvements that could be added to the shared styles?
https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/components/modal_component/modal_component.scss

File: spec/system/admin/order_cycles/edit_spec.rb [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    it "properly toggles order cycle save bar buttons to show warning modal" do
+      create(:schedule, name: 'Schedule1', order_cycles: [oc0])
+
+      # When I go to the admin order cycle edit page
+      login_as_admin
+      visit edit_admin_order_cycle_path(oc0)
+
+      expect(page).to have_selector("#linked-schedule-warning-modal")
+      expect(page).not_to have_selector("#modal-actions")
+      expect(page).to have_selector("#form-actions")
```
Comment (dacook) [inline spec/system/admin/order_cycles/edit_spec.rb:29]: Where possible, we prefer to check for things that the user can see.
In this case, I think these expectations ☝️  aren't necessary.

File: spec/system/admin/order_cycles/edit_spec.rb [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+      # Enable savebar save buttons to open warning modal
+      expect(page.find('#order_cycle_orders_close_at').value).to eq time.strftime('%Y-%m-%d %H:%M')
+      expect(page).not_to have_selector("#form-actions")
+      expect(page).to have_selector("#modal-actions")
+      expect(page).to have_content('You have unsaved changes')
+      expect(page).not_to have_selector('#linked-schedule-warning-modal .reveal-modal.in')
+
+      # click save to open warning modal
+      click_button('Save')
+      expect(page).to have_selector('#linked-schedule-warning-modal .reveal-modal.in')
```
Comment (dacook) [inline spec/system/admin/order_cycles/edit_spec.rb:52]: Here, we can check that the warning message is readable by the user, ie:

```suggestion
      expect(page).to have_content "Orders are linked to this cycle"
```

File: app/views/admin/order_cycles/edit.html.haml (hunk 4 above)
Comment (dacook) [reply app/views/admin/order_cycles/edit.html.haml:58]: I agree, it doesn't look right, but not sure of a better way.  At least it's a way to continue the transition to Stimulus, away from Angular.

**Review by dacook — CHANGES_REQUESTED** [review]
Thanks for working on this, it looks like a big task!

In addition to Gaetan's comments, I note in the issue that it depends if any orders are made in the order cycle, it's not directly about "schedules" which are a different concept, so I think that will need changing.

Secondly, I think there's two ways to solve it. We could validate on the client-side (as you have), or on the server-side, then trigger the modal. I think server-side might have been simpler, and it has the advantage of checking up-to-date information from the database (for example an order could have been made after the page loaded).
Still, there's advantages to validating client-side, it's nice and quick.

I have a couple of suggestions to simplify it though, which I hope will make the code easier to maintain in the long run. Could you please have a look?

**Comment by wandji20** [comment]
Hi
I decided to give a shot with validations on the server side which seems less complicated and clunky.

**Comment by dacook** [comment]
Thanks sorry I wasn't able to review this week; we will review on Monday.

File: app/controllers/admin/order_cycles_controller.rb [hunk 9]
```diff
@@ -299,5 +295,23 @@ def order_cycle_bulk_params
         collection_attributes: [:id] + PermittedAttributes::OrderCycle.basic_attributes
       ).to_h.with_indifferent_access
     end
+
+    # Check that order cycle datetime values changed if it has existing orders
+    def verify_datetime_change
+      return unless params[:order_cycle][:confirm]
+      return unless @order_cycle.orders.exists?
+      return if same_dates(@order_cycle.orders_open_at&.to_s,
```
Comment (rioug) [inline app/controllers/admin/order_cycles_controller.rb:303]: `@order_cycle.orders_open_at` is `ActiveSupport::TimeWithZone` so can just do `@order_cycle.orders_open_at.to_datetime`  , no need to cast to string to then re parse it.
Same with `@order_cycle.orders_close_at` below.

Could you rewrite this to remove the unnecessary casting/parsing ?

File: app/assets/javascripts/admin/order_cycles/services/order_cycle.js.coffee [hunk 10]
```diff
@@ -161,14 +161,25 @@ angular.module('admin.orderCycles').factory 'OrderCycle', ($resource, $window, $
...
+        })
       oc.$update {order_cycle_id: this.order_cycle.id, reloading: (if destination? then 1 else 0)}, (data) =>
+        # Hide all confirmation buttons in warning modal
+        $('#linked-order-warning-modal .modal-actions button.secondary').css({ display: 'none' })
+        # Show the appropriate confirmation button, open warning modal, and return
+        if data.trigger_action
+          StatusMessage.display 'notice', "You have unsaved changes"
+          $("#linked-order-warning-modal button[data-trigger-action=#{data.trigger_action}]").css({ display: 'block' });
+          $('.warning-modal button.modal-target-trigger').trigger('click');
+          return;
```
Comment (rioug) [inline app/assets/javascripts/admin/order_cycles/services/order_cycle.js.coffee:177]: Usually it's not great to be calling Jquery directly in angular, but here we are trying to update something that is controlled by `stimulusjs` and I can't really think of any other way we could do this ? 

We could probably have a stimulus controller handle the show/hide button with some eventListener on the button which would react to certain element attribute change, then we could update the element attribute via angular/jquery.  But I don't think that's worth the hassle, given the existing solution is easy to understand.

File: app/assets/javascripts/admin/order_cycles/services/order_cycle.js.coffee [hunk 11]
```diff
@@ -161,8 +161,21 @@ angular.module('admin.orderCycles').factory 'OrderCycle', ($resource, $window, $
...
```
[the remark this answers is not in the document]
Comment (dacook) [reply app/assets/javascripts/admin/order_cycles/services/order_cycle.js.coffee:177]: I have to confess I still don't see what this is for.. could you please explain?

**Review by dacook — COMMENTED** [review]
Thanks Wandji, I also have some comments, and request for some explanation

File: app/assets/javascripts/admin/order_cycles/services/order_cycle.js.coffee (hunk 11 above)
[the remark this answers is not in the document]
Comment (wandji20) [reply app/assets/javascripts/admin/order_cycles/services/order_cycle.js.coffee:177]: Sure @dacook 
So what happens here is that I

- Set the status display message to show unsaved changes
 Why? 
 Before the request is made, we set the status message to **Saving**, and we still get the **Saving** status message after the warning modal is opened.
 And I agree, this should be translated.
 
 Then find and display the appropriate conform button from the warning modal.
 
 Also, I will move the modal confirmation button-related logic to a stimulus controller.

File: app/views/admin/order_cycles/_date_time_warning_modal_content.html.haml [hunk 12]
```diff
@@ -0,0 +1,14 @@
+.modal-body.flex-column-gap-1
```
[the remark this answers is not in the document]
Comment (wandji20) [reply app/views/admin/order_cycles/_date_time_warning_modal_content.html.haml:1]: Thanks for noticing this one
I am using` .flex-column.gap-1` so will remove `.modal-body`

File: app/models/order_cycle.rb [hunk 13]
```diff
@@ -24,6 +24,7 @@ class OrderCycle < ApplicationRecord
                                          where incoming: false
                                        }, class_name: "Exchange", dependent: :destroy
 
+  has_many :orders, class_name: 'Spree::Order', dependent: :restrict_with_exception
```
[the remark this answers is not in the document]
Comment (wandji20) [reply app/models/order_cycle.rb:27]: Adding `dependent: :restrict_with_exception` here was to help handle rubocop error (Rails/HasManyOrHasOneDependent: Specify a `:dependent` option.)
I thought of adding to `.rubocop_todo.yml` but felt this was better.

[the remark this answers is not in the document]
Comment (dacook) [reply app/models/order_cycle.rb:27]: I see. Yes I agree this was certainly better 👍

File: app/controllers/admin/order_cycles_controller.rb [hunk 14]
```diff
@@ -300,18 +300,18 @@ def order_cycle_bulk_params
...
 
-    def same_dates(string1, string2)
-      false unless string1 && string2
+    def same_dates(date, string)
+      false unless date && string
 
-      DateTime.parse(string1).strftime('%Y-%m-%d %H:%M') ==
-        DateTime.parse(string2).strftime('%Y-%m-%d %H:%M')
+      DateTime.parse(string).strftime('%Y-%m-%d %H:%M') ==
+        date.to_datetime.strftime('%Y-%m-%d %H:%M')
```
Comment (rioug) [inline app/controllers/admin/order_cycles_controller.rb:314]: I couldn't understand why you needed to format the dates to compare them, but then I realised we only need to compare up to the minute and we don't care about the seconds. It's fine like this, but you could also use `to_fs()` with a predefine format ie: 
```suggestion
      DateTime.parse(string).to_fs(:short) == date.to_fs(:short)
```

**Review by rioug — CHANGES_REQUESTED** [review]
Nice one, I like how you handle the communication between angular and stimulus js via the html attributes. Last thing, could you remove `jQuery` from the stimulus controller ? Thanks :pray:

**Review by rioug — APPROVED** [review]
Awesome ! thank you for working with us through this one :pray:

File: spec/system/admin/order_cycles/edit_spec.rb [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+      click_button('Save')
+      expect(page).not_to have_content "Orders are linked to this order cycle"
+      expect(page).to have_content('Your order cycle has been updated.')
+
+      # change date range field value
+      time = DateTime.current
+      find('#order_cycle_orders_close_at').click
+      select_datetime_from_datepicker Time.zone.at(time)
+
+      expect(page.find('#order_cycle_orders_close_at').value).to eq time.strftime('%Y-%m-%d %H:%M')
```
Comment (dacook) [inline spec/system/admin/order_cycles/edit_spec.rb:33]: This spec doesn't specify what the close_at time _was_ or is changed _to_. 
Apparently they are different in this context, but it's better to make that explicit, for example:
* Create an order cycle with times specified (eg `create(:simple_order_cycle, orders_close_at: `"2024-08-08 00:00"`)
* Create the order for that cycle.
* Specify a time for the selection: `select_datetime_from_datepicker Time.zone.parse("2024-08-09 00:00")`
* Repeat the time for the expect: `expect(page.find('#order_cycle_orders_close_at').value).to eq "2024-08-09 00:00"`

Having the date/times written in the code makes it easier to read and understand the spec, which saves time and helps us avoid mistakes.

**Review by dacook — CHANGES_REQUESTED** [review]
Thanks for persisting Wandji. It's unfortunate that we need such a complicated solution, but we're getting there. I'm a uneasy using MutationObserver to watch for the modals, which seems brittle and likely to break. For example, what if a DOM ID or class change in the future? But at least it's covered well by a spec.

However I noticed that there are two angular controllers updated (edit and simple_edit), so I guess there's two contexts that this modal confirmation can occur? If so, then we should have a spec for both. Would you be able to add a spec for the other context?

PS while we're looking at specs, I have a suggestion to update it also 👇

File: spec/system/admin/order_cycles/edit_spec.rb [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+        expect(page).to have_field 'order_cycle_name', with: "Order cycle name updated"
+        expect(page).to have_content('Your order cycle has been updated.')
+
+        # change date range field value
+        find('#order_cycle_orders_close_at').click
+        within(".flatpickr-calendar.open") do
+          expect(page).to have_selector '.shortcut-buttons-flatpickr-buttons'
+          select_datetime_from_datepicker Time.zone.parse("2024-03-30 00:00")
+          find("button", text: "Close").click
+        end
```
Comment (dacook) [inline spec/system/admin/order_cycles/edit_spec.rb:48]: I'm surprised to see that we don't have a helper method to handle this, it seems like a common action.

**Review by dacook — APPROVED** [review]
Thanks, that looks like it covers it well. 
I can see that we already had spec files that cover simple and complex order cycles maybe you could have added to those. 
But I guess it also makes sense to group these similar specs together in one file like you have. 🤷

**Comment by drummer83** [comment]
Hi @wandji20,
Thank you for this PR! Here are my testing notes.

## Test cases

The popup: :heavy_check_mark: 
image

- When changing the date range of an OC which has linked orders, display a popup, for instance: "Orders are linked to this order cycle. If you wish to create a new order cycle, it is recommended to duplicate the order cycle first and then change the dates." :heavy_check_mark: 
- The popup is independent from added schedules. :heavy_check_mark: 
- The popup is shown when clicking the 'save', 'save and next' (complex order cycle) and 'save and back to list' (simple order cycle) buttons. :heavy_check_mark: 
- The popup is shown when changing start date only, end date only, or both. :heavy_check_mark: 
- The popup is shown on the simple and on the complex order cycle editing page. :heavy_check_mark: 
- The popup is shown for admin and super admin users. :heavy_check_mark: 
- The popup is shown for past, active, and future order cycles. :heavy_check_mark: 
- The popup is only shown for order cycles that have orders linked to it. :heavy_check_mark: 
- The popup has two buttons: "Proceed anyway" & "Cancel" :heavy_check_mark: 
  - Proceed anyway => the saving process goes according to what the user has selected (saving and staying on the form, or saving and going to the next page, or saving and going back to the list). :heavy_check_mark: 
  - Cancel should move the user back to the list (where the duplicate button is visible) :x: 
    The popup is closed, but the user stays on the same page and the changed information is still there.

## Notes
- The popup only works on the edit page of the OC (as specified), but dates can be adjusted in the OC list as well. :question: 
- Warning is displayed as soon as an order is in 'cart' state, it's not waiting for 'complete'. :question: 

## Conclusion
Everything is working well! :muscle: 
I only found one mismatch between the specification and the implementation, which is the behaviour after clicking 'Cancel'. Simply closing itself is what usually happens to a popup when clicking cancel - so the implementation is according to some kind of standard. However, the user has to make the additional click on the cancel button on the order cycles edit page to return to the list of order cycles, all other changes will be lost.
@openfoodfoundation/train-drivers-product-owners Can you please comment whether we can move forward as is, whether we need an adjustment of the implementation, or open a follow-up issue. Thanks!
@openfoodfoundation/train-drivers-product-owners Can you also comment on the findings in the 'Notes' section, please. Thanks!
Adding the feedback-needed label for now.

Thanks again! :pray:

**Comment by wandji20** [comment]
Thanks for testing @drummer83 
Oops, I completely missed the cancel button interaction.
I am unsure how it works from here, but I would not mind addressing these issues in another PR.

**Comment by drummer83** [comment]
Let's see what @RachL 's opinion is on this. :pray:

**Comment by RachL** [comment]
Apologies for the late answer here, thanks Konrad for the reminder.

thanks @wandji20 for working on this one, it will be life changing for some users!

And thanks @drummer83 for the detailed testing, as always 😍 

My proposal would go for merging as is and handling : 

- behavior of cancel button

- displaying the popup on the OC list

in one or 2 follow-up PRs / issues. @wandji20 would you be available to work on both? maybe the OC list should be more a priority than the cancel button IMO.

Regarding the fact that the order is in cart state and not complete state, I see an advantage: at least if for whatever reason the user has several OC open at the same time, maybe orders in cart state can be completed 🤷‍♀️  Maybe this is an edge case, so anyway let's see with user feedback how to proceed on this one.

any thoughts?

**Comment by drummer83** [comment]
I agree, the topic around the order state 'cart' is an edge case. No need for action here, I think.
I will open issues for the two other topics remaining.
Merging! :boom:

**Comment by drummer83** [comment]
Created #12774 and #12775 to follow up.
